### ISSUE TITLE
Parallelize Secure Random operation

### ISSUE THREAD

**Issue by justin1121** [mortendahl/tf-encrypted#270]
Now that we (almost) have a function to generate secure random numbers, we need to make sure it generates the numbers in parallel. This will help generate large tensors full of random numbers even faster than before. The basics of this is that we'll use the `Shard` function provided in the tensorflow c++ api to split the work up among threads. There are some finicky details surrounding parallelization that will need to be figured out once the PR is merged. (PR: https://github.com/mortendahl/tf-encrypted/pull/265)

Note: In the latest version of tensorflow the `Shard` function is deprecated so this may be removed at some point in the future.

Example of `Shard` usage: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/random_op.cc#L167

**Comment by justin1121** [comment]
With https://github.com/mortendahl/tf-encrypted/pull/347/ merged we have parallelized random_uniform but the seeded version is not yet parallelized. It is a little more complicated because many threads need to read the cipher stream in the correct order and place them into the output buffer in the correct order so that multiple machines generate the same output for the same seed. See the above comment for some more details. I had one implementation of this but it required two copies and was not any faster than the version that was not parallelized. 

I have some ideas on how we do this by assigning one cipher stream to each thread using a different nonce for each thread but the number of shards will need to be consistent between machine. This is doable using a new shard function in tensorflow called `TransformRangeConcurrently` but only exists in tf 1.12.0. Will need to discuss with @mortendahl to figure out some of the finer details surrounding this.

**Comment by mortendahl** [comment]
Great write up @justin1121! I think we should avoid any kind of constraints on the number of threads, but we can do much the same by simply putting the constraints on number of work units instead as in work stealing; is this what you mean by `TransformRangeConcurrently`?

**Comment by justin1121** [comment]
Yep. `TransformRangeConcurrently` has a predictable number of work units based on the inputs. We can come up with some heuristic to calculate the 'block size' which is the main factor in how many work units are created. This way each machine will always generate the same number of work units.
